File: src/main/java/wraith/fabricaeexnihilo/api/recipes/witchwater/WitchWaterEntityRecipe.java
package wraith.fabricaeexnihilo.api.recipes.witchwater;

import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityType;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.passive.VillagerEntity;
import net.minecraft.village.VillagerProfession;
import wraith.fabricaeexnihilo.api.crafting.EntityTypeIngredient;

import java.util.function.Predicate;

public record WitchWaterEntityRecipe(EntityTypeIngredient target,
                                     VillagerProfession profession,
                                     EntityType<?> toSpawn) implements Predicate<Entity> {

    public boolean test(Entity entity) {
        if (!(entity instanceof LivingEntity)) {
            return false;
        }
        if (target.test(entity.getType())) {
            return entity instanceof VillagerEntity villager ? villager.getVillagerData().getProfession() == profession : profession == null;
        }
        return false;
    }

}

File: src/main/kotlin/exnihilofabrico/api/recipes/WitchWaterEntityRecipe.kt
package exnihilofabrico.api.recipes

import exnihilofabrico.api.crafting.EntityTypeIngredient
import net.minecraft.entity.Entity
import net.minecraft.entity.EntityType
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.passive.VillagerEntity
import net.minecraft.village.VillagerProfession
import java.util.function.Predicate

data class WitchWaterEntityRecipe(val target: EntityTypeIngredient,
                                  val profession: VillagerProfession?,
                                  val tospawn: EntityType<*>
): Predicate<Entity?> {
    override fun test(entity: Entity?): Boolean {
        if(entity == null || entity !is LivingEntity)
            return false
        if(target.test(entity.type)) {
            if(entity is VillagerEntity) {
                return entity.villagerData.profession == profession
            }
            else
                return profession == null

        }
        return false
    }
}